scalpel-core/src/Text/HTML/Scalpel/Internal/Scrape.hs

```diff
@@ -31,54 +38,39 @@ import qualified Text.StringLike as TagSoup
 
 -- | A value of 'Scraper' @a@ defines a web scraper that is capable of consuming
 -- a list of 'TagSoup.Tag's and optionally producing a value of type @a@.
-newtype Scraper str a = MkScraper {
-        scrapeTagSpec :: TagSpec str -> Maybe a
-    }
-
-instance Functor (Scraper str) where
-    fmap f (MkScraper a) = MkScraper $ fmap (fmap f) a
-
-instance Applicative (Scraper str) where
-    pure = MkScraper . const . Just
-    (MkScraper f) <*> (MkScraper a) = MkScraper applied
-        where applied tags | (Just aVal) <- a tags = ($ aVal) <$> f tags
-                           | otherwise             = Nothing
-
-instance Alternative (Scraper str) where
-    empty = MkScraper $ const Nothing
-    (MkScraper a) <|> (MkScraper b) = MkScraper choice
-        where choice tags | (Just aVal) <- a tags = Just aVal
-                          | otherwise             = b tags
-
-instance Monad (Scraper str) where
-    fail = Fail.fail
-    return = pure
-    (MkScraper a) >>= f = MkScraper combined
-        where combined tags | (Just aVal) <- a tags = let (MkScraper b) = f aVal
-                                                      in  b tags
-                            | otherwise             = Nothing
-
-instance MonadPlus (Scraper str) where
-    mzero = empty
-    mplus = (<|>)
-
-instance Fail.MonadFail (Scraper str) where
-    fail _ = mzero
-
--- | The 'scrape' function executes a 'Scraper' on a list of
+newtype ScraperT str m a = MkScraper (ReaderT (TagSpec str) (MaybeT m) a)
+    deriving (Functor, Applicative, Alternative, Monad, MonadPlus, Fail.MonadFail)
+
+instance MonadTrans (ScraperT str) where
+  lift op = MkScraper . lift . lift $ op
+
+instance MonadIO m => MonadIO (ScraperT str m) where
+  liftIO = lift . liftIO
+
+type Scraper str = ScraperT str Identity
```

Comment:
Please restore the haddock comments for this type.

In general, I think having the "standard" comments on the simpler type would be easier for folks who aren't familiar with monad transformers. Then for the `*T` types and functions we can document how they differ from the non-transformer version.